internal/build/dockerfile.go

```diff
@@ -47,15 +55,36 @@ func (d Dockerfile) RmPaths(pathsToRm []pathMapping) Dockerfile {
 	return d.join(newDf)
 }
 
-func (d Dockerfile) ForbidEntrypoint() error {
-	for _, line := range strings.Split(string(d), "\n") {
-		if strings.HasPrefix(line, "ENTRYPOINT") {
+func (d Dockerfile) ValidateBaseDockerfile() error {
+	result, err := parser.Parse(bytes.NewBufferString(string(d)))
+	if err != nil {
+		return fmt.Errorf("ValidateBaseDockerfile: %v", err)
+	}
+
+	err = traverse(result.AST, func(node *parser.Node) error {
```

Comment:
is this markedly more efficient than doing it the way we did it at left? why bother with this kind of traversal?

so the risk here is that we flag any line that contains the word "add", "copy", "cmd", etc even if it's not Dockerfile syntax, even if it's a comment, etc. -- which is why at left i checked the beginnings of lines. idk if it's worth worrying about now, but something to note.

Reply:
(or does `parser.Parse` split stuff out in such a way that checking `result.AST` is equivalent to checking the docker verb?)

Reply:
i added some tests for cases that you're worried about, both cases where a hand-rolled regexp-based parser would fall down, and cases where you're worried about 'add' in a command.

yes, the AST is structured in a way so that each command is an AST node.